lib/rucio/client/uploadclient.py

```diff
@@ -153,11 +156,23 @@ def upload(self, items, summary_file_path=None):
                 logger.warning('Upload with given pfn implies that no_register is True, except non-deterministic RSEs')
                 no_register = True
 
-            if not no_register:
+            if not no_register and not register_after_upload:
                 self._register_file(file, registered_dataset_dids)
 
-            # if file already exists on RSE we're done
-            if not is_deterministic and not no_register:
+            # if register_after_upload, file should be overwritten if it is not registered
+            # otherwise if file already exists on RSE we're done
+            if register_after_upload:
+                if rsemgr.exists(rse_settings, file_did):
+                    try:
+                        dids = self.client.get_did(file['did_scope'], file['did_name'])
+                        logger.info('File already registered. Skipping upload.')
+                        logger.debug(dids)
+                        continue
+                    except DataIdentifierNotFound:
+                        logger.info('File already exists on RSE. Existing replica will be removed first.')
+                        rsemgr.delete(rse_settings, [{'name': file['did_name'], 'scope': file['did_scope']}])
```

Comment:
I don't think this workflow is correct.
rsemgr.upload(...) already takes care of the proper deletion if an incomplete file got uploaded (Check rsemanager.upload and the protocols). So you don't have to do this here. I think it might be enough to just call the upload and catch the exception and then do the registration